File: botreq-status/botreq-status.rb
require 'io/console'
require 'mediawiki/butt'
require 'time'
require_relative 'marshallable-butt.rb'

LOGIN_STATE_FILE = 'login_state'
USERNAME = 'APersonBot'
BOTREQ = 'Wikipedia:Bot requests'
BOTOP_CAT = 'Category:Wikipedia bot owners'
REPORT_PAGE = 'User:APersonBot/BOTREQ status'

HEADER = /== ?([^=]+) ?==\n/#/== ?([\w\[\]:\{\},\. ]+) ?==\n/
USER = /\[\[User.*?:(.*?)(?:\||(?:\]\]))/
TIMESTAMP = /\d{2}:\d{2}, \d{1,2} [A-Za-z]* \d{4}/
SIGNATURE = /\[\[User.*?\]\].*?\(UTC\)/

SECONDS_IN_DAY = 24 * 60 * 60
OLD_REQUEST_TIME = 60 * SECONDS_IN_DAY # The number of seconds after which a request is old
TIME_FORMAT_STRING = '%Y-%m-%d, %H:%M'

# Log in
$wiki = nil
if File.exists? LOGIN_STATE_FILE
  puts "Attempting to read login state from #{LOGIN_STATE_FILE}..."
  $wiki = Marshal.load(File.read(LOGIN_STATE_FILE))
  puts $wiki
  puts $wiki.user_bot?
end
if !$wiki || !$wiki.user_bot?
  $wiki = MediaWiki::Butt.new('https://en.wikipedia.org/w/api.php')
  print "Password for #{USERNAME} on en.wikipedia.org: "
  password = STDIN.noecho(&:gets).chomp
  puts ''
  $wiki.login(USERNAME, password)
end
puts "Logged in as #{$wiki.get_current_user_name} (bot: #{$wiki.user_bot?})"

# Set up a couple of classes
class Request
  attr_accessor :title, :replies, :last_editor, :last_edit_time, :last_botop_editor, :last_botop_time

  # Writes out a wikitext row for use in the BOTREQ status page.
  def row
    replies_style = 'style="background: red;" | ' if @replies == 0

    title = @title.gsub(/\[\[(?:.+\|)?(.+)\]\]/, '\1')
    anchor = title.gsub(/{{tl\|(.+)}}/, '.7B.7B\1.7D.7D')
    title = title.gsub(/{{tl\|(.+)}}/, '\1')

    last_edit_time_style = 'style="background: red;" | ' if (Time.now - @last_edit_time) > OLD_REQUEST_TIME

    @last_botop_time = @last_botop_time ? @last_botop_time.strftime(TIME_FORMAT_STRING) : '{{n/a}}'

    %Q{
|-
| [[Wikipedia:Bot requests##{title}|#{title}]] || #{replies_style}#{@replies} || #{@last_editor} || #{last_edit_time_style}#{@last_edit_time.strftime(TIME_FORMAT_STRING)} || #{@last_botop_editor} || #{@last_botop_time}}
  end
end

class BotopChecker
  attr_accessor :results

  def check(user)
    if results.include? user
      results[user]
    else
      categories = $wiki.get_categories_in_page("User:#{user}")
      results[user] = if categories.nil? || categories.length == 0
                        false
                      else
                        categories.include? BOTOP_CAT
                      end
    end
  end
end

# We want a hash of the form {section_title => section_body}
page_content = $wiki.get_text(BOTREQ)
page_content = "==" + page_content.split("==", 2)[1]
section_matches = page_content.to_enum(:scan, HEADER).map { Regexp.last_match }
headings = section_matches.map { |x| x[1].strip }
body_offsets = section_matches.map { |x| x.offset(0) }.flatten[1..-1].each_slice(2)
bodies = body_offsets.map { |x| page_content[x[0]..(x.length == 1 ? -1 : x[1] - 1)] }
sections = Hash[headings.zip(bodies)]
puts "#{sections.length} sections found."

# Build request objects
botop_checker = BotopChecker.new
botop_checker.results = {}
requests = sections.map do |heading, body|
  signatures = body.to_enum(:scan, SIGNATURE)
               .map { Regexp.last_match }
               .map { |x| x.to_s }
               .map { |match| [USER.match(match)[1], TIMESTAMP.match(match)[0]] }
               .select { |user, timestamp| !user.include? '/' }
  request = Request.new
  request.title = heading
  request.replies = body.scan(/\(UTC\)/).length - 1
  request.last_editor = signatures[-1][0]
  request.last_edit_time = Time.parse(signatures[-1][1])
  last_botop_signature = signatures.reverse.find { |user, timestamp| botop_checker.check(user) }
  if last_botop_signature
    request.last_botop_editor, request.last_botop_time = last_botop_signature
    request.last_botop_time = Time.parse(request.last_botop_time)
  else
    request.last_botop_editor = '{{no result|None}}'
    request.last_botop_time = nil
  end
  request
end
header = %({| border="1" class="sortable wikitable plainlinks"
!Title !! Replies !! Last editor !! Date/Time !! Last botop editor !! Date/Time
)
final_text = header + requests.map { |x| x.row }.join('') + "\n|}"

puts "Text generated. Saving to #{REPORT_PAGE}..."

print 'Press enter to continue...'
gets

result = $wiki.edit(REPORT_PAGE, final_text, false, false, 'Generating a statusreport for BOTREQ')
print 'Result: '
puts result

File: botreq-status/botreq-status.py
import datetime
import itertools
import mwparserfromhell
import pywikibot
import re
import sys
import urllib

BOTREQ = "Wikipedia:Bot requests"
BOTREQ_HTML_URL = "https://en.wikipedia.org/w/index.php?title=Wikipedia:Bot_requests&action=view"
BOTOP_CAT = "Wikipedia bot operators"
REPORT_PAGE = "User:EnterpriseyBot/BOTREQ status"
TABLE_HEADER = """<noinclude>{{botnav}}This is a table of current [[WP:BOTREQ|]] discussions, updated automatically by {{user|EnterpriseyBot}}.</noinclude>
{| border="1" class="sortable wikitable plainlinks"
! # !! Title !! Replies !! Last editor !! Date/Time !! Last botop editor !! Date/Time
"""
SUMMARY = "Bot updating BOTREQ status table ({} requests)"
USER_NONE_WIKITEXT = "{{sort|ω|{{no result|None}}}}"

USER = re.compile(r"\[\[User.*?:(.*?)(?:\||(?:\]\]))")
TIMESTAMP = re.compile(r"\d{2}:\d{2}, \d{1,2} [A-Za-z]* \d{4}")
SIGNATURE = re.compile(r"\[\[User.*?\]\].*?\(UTC\)")
SECTION_HEADER = re.compile(r"^==([^=]|\s+).*?\s*==$", flags=re.M)

SIGNATURE_TIME_FORMAT = "%H:%M, %d %B %Y"
TIME_FORMAT_STRING = "%Y-%m-%d, %H:%M"

class Request:
    pass

def make_table_row(r):
    if type(r.last_edit_time) is datetime.datetime:
        old = (datetime.datetime.now() - r.last_edit_time).days > 60
        r.last_edit_time = ('style="background: red;" | ' if old else '') + r.last_edit_time.strftime(TIME_FORMAT_STRING)

    if type(r.last_botop_time) is datetime.datetime:
        r.last_botop_time = r.last_botop_time.strftime(TIME_FORMAT_STRING)

    # Add a red backgroud to the replies
    replies = ('style="background: red;" | ' if r.replies == 0 else '') + str(r.replies)

    def user_link(username):
        return USER_NONE_WIKITEXT if username == USER_NONE_WIKITEXT else '[[User:' + username + '|' + username + ']]'

    elements = map(str, [r.row_number, r.html_id, r.title, replies, user_link(r.last_editor), r.last_edit_time, user_link(r.last_botop_editor), r.last_botop_time])
    return u"|-\n| {} || [[WP:Bot requests#{}|{}]] || {} || {} || {} || {} || {}".format(*elements)

botop_cache = {}
def is_botop(wiki, username):
    if username in botop_cache:
        return botop_cache[username]

    userpage = pywikibot.Page(wiki, "User:" + username)
    result = any(x.title(with_ns=False) == BOTOP_CAT for x in userpage.categories())
    botop_cache[username] = result
    return result

def get_section_titles_and_ids():
    html = urllib.request.urlopen(BOTREQ_HTML_URL).read().decode('utf-8')
    sections = []
    for matchobj in re.finditer(r'<h2.+?<span class="mw-headline" id="([^"]+)".+?</h2>', html):
        # sorry this is awful
        title = re.sub('<a .+?>(.+?)</a>', '\\1', re.sub('^(<.+?>)+', '', re.sub('<span.+?</span>', '', re.sub('<a (?!href="[^"]).+?</a>', '', matchobj.group(0)))).partition('</span>')[0])
        id = matchobj.group(1)
        sections.append({'title': title, 'id': id})
    return sections

def main():
    wiki = pywikibot.Site("en", "wikipedia")
    wiki.login()
    botreq = pywikibot.Page(wiki, BOTREQ)
    page_content = botreq.text

    section_headers = list(SECTION_HEADER.finditer(page_content))

    # If it's not a level-2 header, the char before a match will be "="
    section_headers = list(filter(lambda h:page_content[h.start(0) - 1] != "=",
                             section_headers))

    # Now, build our list of sections
    sections = []
    for i, section_header_match in enumerate(section_headers):
        if i + 1 < len(section_headers):
            next_section_header = section_headers[i + 1]
            next_section_start = next_section_header.start(0)
        else:
            next_section_start = len(page_content) + 1
        this_section_end = next_section_start - 1
        this_section_start = section_header_match.end(0)
        section_content = page_content[this_section_start:this_section_end]
        section_content = section_content.strip()
        sections.append(section_content)

    def section_to_request(enumerated_section_tuple):
        enum_number, section_wikitext = enumerated_section_tuple
        section = mwparserfromhell.parse(section_wikitext)
        r = Request()
        r.row_number = enum_number + 1
        r.replies = section.count(u"(UTC)") - 1
        signatures = []
        for index, each_node in enumerate(section.nodes):
            if type(each_node) == mwparserfromhell.nodes.text.Text and "(UTC)" in each_node:

                # Get the last timestamp-looking thing (trick from http://stackoverflow.com/a/2988680/1757964)
                each_node = str(each_node)
                for timestamp_match in TIMESTAMP.finditer(each_node): pass
                try:
                    timestamp = datetime.datetime.strptime(timestamp_match.group(0), SIGNATURE_TIME_FORMAT)
                except ValueError:
                    timestamp = "{{unknown}}"

                # Use the last user talk page link before the timestamp
                for user_index in itertools.count(index - 1, -1):
                    user = USER.search(str(section.get(user_index)))
                    if user:
                        user = user.group(1)
                        break

                # Check for user renames/redirects
                user_page = pywikibot.Page(wiki, "User:" + user)
                if user_page.isRedirectPage():
                    redirect_text = user_page.get(get_redirect=True)
                    user_wikicode = mwparserfromhell.parse(redirect_text)
                    redirect_link = user_wikicode.filter_wikilinks()[0]
                    user = redirect_link.title.split(":")[1]

                signatures.append((user, timestamp))
        # Process usernames by removing anchors
        signatures = [(x.partition('#')[0], y) for x, y in signatures]

        # Default values for everything
        r.last_editor, r.last_edit_time = r.last_botop_editor, r.last_botop_time = USER_NONE_WIKITEXT, "{{n/a}}"

        if signatures:
            r.last_editor, r.last_edit_time = signatures[-1]
            for user, timestamp in reversed(signatures):
                if is_botop(wiki, user):
                    r.last_botop_editor, r.last_botop_time = user, timestamp
                    break
        return r

    # Why enumerate? Because we need row numbers in the table
    requests = list(map(section_to_request, enumerate(sections)))

    # Add in title & HTML id
    section_titles_and_ids = get_section_titles_and_ids()
    if len(requests) != len(section_titles_and_ids):
        raise Exception('len(requests) != len(section_titles_and_ids): {} != {}'.format(len(requests), len(section_titles_and_ids)))
    for (request, title_and_id) in zip(requests, section_titles_and_ids):
        request.title = title_and_id['title']
        request.html_id = title_and_id['id']

    table_rows = map(make_table_row, requests)
    table = "\n".join(table_rows) + "\n|}"
    wikitext = TABLE_HEADER + table

    report_page = pywikibot.Page(wiki, REPORT_PAGE)
    report_page.text = wikitext
    report_page.save(quiet=True, summary=SUMMARY.format(len(list(requests))))

if __name__ == "__main__":
    main()
